code/gf128/Hacl.Spec.GF128.PolyLemmas_helpers.fst

```diff
@@ -0,0 +1,734 @@
+    == {lemma_mod_distribute (monomial 126) (monomial 124) (monomial 64)}
+    (a *. (monomial 126 %. monomial 64 +. monomial 124 %. monomial 64 +. monomial 114 %. monomial 64)) %. monomial 64;
+    == {
+      lemma_mul_monomials 62 64;
+      lemma_add_zero (monomial 126);
+      lemma_div_mod_unique ((monomial 126) +. zero) (monomial 64) (monomial 62) zero;
+      // --> monomial 126 %. monomial 64 == zero
+      lemma_mul_monomials 60 64;
+      lemma_add_zero (monomial 124);
+      lemma_div_mod_unique ((monomial 124) +. zero) (monomial 64) (monomial 60) zero;
+      // --> monomial 126 %. monomial 64 == zero
+      lemma_mul_monomials 50 64;
+      lemma_add_zero (monomial 114);
+      lemma_div_mod_unique ((monomial 114) +. zero) (monomial 64) (monomial 50) zero
+      // --> monomial 126 %. monomial 64 == zero
+    }
+    (a *. (zero +. zero +. zero)) %. monomial 64;
+    == {lemma_add_zero zero}
+    (a *. zero) %. monomial 64;
+    == {lemma_mul_zero a}
+    zero;
+  }
+
+let lemma_shift_left_cancel a a0 a1 =
+  reveal_defs ();
+  calc (==) {
+    shift (((shift a1 64) +. a0) %. monomial 64) 64;
+    == {lemma_shift_is_mul a1 64}
+    shift ((a1 *. monomial 64 +. a0) %. monomial 64) 64;
+    == {lemma_mod_distribute (a1 *. monomial 64) a0 (monomial 64)}
+    shift ((a1 *. monomial 64) %. monomial 64 +. a0 %. monomial 64) 64;
+    == {lemma_mod_small a0 (monomial 64)}
+    shift ((a1 *. monomial 64) %. monomial 64 +. a0) 64;
+    == {lemma_mod_mul_mod_right a1 (monomial 64) (monomial 64)}
+    shift ((a1 *. (mod (monomial 64) (monomial 64))) %. monomial 64 +. a0) 64;
+    == {lemma_mod_cancel (monomial 64)}
+    shift ((a1 *. zero) %. monomial 64 +. a0) 64;
+    == {lemma_mul_zero a1; lemma_add_zero_left a0}
+    shift a0 64;
+  }
+
+let lemma_shift_right_cancel a a0 a1 =
+  lemma_shift_is_div a 64;
+  lemma_shift_is_mul a1 64;
+  lemma_div_mod_unique a (monomial 64) a1 a0
+
+let lemma_add_left_shift a0 a1 b =
+  lemma_bitwise_all ();
+  calc (==) {
+    (shift a1 64) +. a0 +. (shift b 64);
+    == {lemma_add_commute (shift a1 64) a0}
+    a0 +. (shift a1 64) +. (shift b 64);
+    == {lemma_add_associate a0 (shift a1 64) (shift b 64)}
+    a0 +. ((shift a1 64) +. (shift b 64));
+    == {lemma_shift_is_mul a1 64; lemma_shift_is_mul b 64}
+    a0 +. (a1 *. monomial 64 +. b *. monomial 64);
+    == {lemma_mul_distribute_left a1 b (monomial 64)}
+    a0 +. (a1 +. b) *. monomial 64;
+    == {lemma_shift_is_mul (a1 +. b) 64}
+    a0 +. (shift (a1 +. b) 64);
+    == {lemma_add_commute a0 (shift (a1 +. b) 64)}
+    (shift (a1 +. b) 64) +. a0;
+  }
+
+let lemma_add_left_shift_double a0 a1 b0 b1 =
+  lemma_bitwise_all ();
+  calc (==) {
+    (shift a1 64) +. a0 +. (shift b1 64 +. b0);
+    == {lemma_add_associate ((shift a1 64) +. a0) (shift b1 64) b0}
+    ((shift a1 64) +. a0 +. shift b1 64) +. b0;
+    == {lemma_add_commute (shift a1 64) a0}
```

Comment:
can't a lot of these be done with the semiring tactic? that sounds very labor intensive doing all of these steps by hand

Reply:
@msprotz Thanks for reviewing the code, I agree that documentation for AES-GCM API should be in-depth and more helpful for the user. I will see how I can put such doc up tomorrow.

Reply:
How can I learn about semiring tactic? or where can I look for example?